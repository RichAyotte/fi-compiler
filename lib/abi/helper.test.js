/**
 * @overview    Helper tests
 */

'use strict'

const fs = require('fs').promises
const {find} = require('lodash')

const compiler = require('../../index')
const helper = require('./helper')
const encodeRawBytes = require('./encodeRawBytes')
const namedTypevalue = require('./namedTypevalue')
const sexp2mic = require('./sexp2mic')

let inputCompiledContract = null
let allCompiledContract = null

beforeAll(async done => {
	const inputTestContract = await fs.readFile('e2e/fi-samples/input_test.fi', {
		encoding: 'utf-8'
	})
	inputCompiledContract = compiler.compile(inputTestContract)

	const allTestContract = await fs.readFile('e2e/fi-samples/all_test.fi', {
		encoding: 'utf-8'
	})
	allCompiledContract = compiler.compile(allTestContract)
	done()
})

describe('ABI Helpers', () => {
	it(`has required helpers`, () => {
		expect(helper).toHaveProperty('pack')
		expect(helper).toHaveProperty('namedTypevalue')
	})

	it(`transforms input to sexp, micheline and raw bytes`, () => {
		const abi = JSON.parse(inputCompiledContract.abi)
		const {input} = find(abi.entry, ['name', 'TestInputSimple'])
		const params = {
			boolLabel: 'true'
			, natLabel: 1234
			, intLabel: -5678
			, stringLabel: `string`
			, mutezLabel: 1000
			, addressLabel: `address`
			, keyLabel: `key`
			, pkhLabel: `pkh`
			, signatureLabel: `signature`
			, bytesLabel: `bytes`
		}

		const sexp = namedTypevalue(input, params, abi)
		expect(sexp).toMatchSnapshot()

		const micheline = sexp2mic(sexp)
		expect(micheline).toMatchSnapshot()

		const rawBytes = encodeRawBytes(micheline)
		expect(rawBytes).toMatchSnapshot()
	})

	it(`transforms all to sexp, micheline and raw bytes`, () => {
		const abi = JSON.parse(allCompiledContract.abi)
		const input = null
		const params = null

		const sexp = namedTypevalue(input, params, abi)
		expect(sexp).toMatchSnapshot()

		const micheline = sexp2mic(sexp)
		expect(micheline).toMatchSnapshot()

		const rawBytes = encodeRawBytes(micheline)
		expect(rawBytes).toMatchSnapshot()
	})
})
