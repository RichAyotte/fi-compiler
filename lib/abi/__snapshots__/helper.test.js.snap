// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABI Helpers transforms all to sexp, micheline and raw bytes 1`] = `"Unit"`;

exports[`ABI Helpers transforms all to sexp, micheline and raw bytes 2`] = `
Object {
  "args": Array [],
  "prim": "Unit",
}
`;

exports[`ABI Helpers transforms all to sexp, micheline and raw bytes 3`] = `"05030b"`;

exports[`ABI Helpers transforms input to sexp, micheline and raw bytes 1`] = `"(Pair True (Pair 1234 (Pair -5678 (Pair \\"string\\" (Pair 1000 (Pair \\"address\\" (Pair \\"key\\" (Pair \\"pkh\\" (Pair \\"signature\\" bytes)))))))))"`;

exports[`ABI Helpers transforms input to sexp, micheline and raw bytes 2`] = `
Object {
  "args": Array [
    Object {
      "args": Array [],
      "prim": "True",
    },
    Object {
      "args": Array [
        Object {
          "int": "1234",
        },
        Object {
          "args": Array [
            Object {
              "int": "-5678",
            },
            Object {
              "args": Array [
                Object {
                  "string": "string",
                },
                Object {
                  "args": Array [
                    Object {
                      "int": "1000",
                    },
                    Object {
                      "args": Array [
                        Object {
                          "string": "address",
                        },
                        Object {
                          "args": Array [
                            Object {
                              "string": "key",
                            },
                            Object {
                              "args": Array [
                                Object {
                                  "string": "pkh",
                                },
                                Object {
                                  "args": Array [
                                    Object {
                                      "string": "signature",
                                    },
                                    Object {
                                      "args": Array [],
                                      "prim": "bytes",
                                    },
                                  ],
                                  "prim": "Pair",
                                },
                              ],
                              "prim": "Pair",
                            },
                          ],
                          "prim": "Pair",
                        },
                      ],
                      "prim": "Pair",
                    },
                  ],
                  "prim": "Pair",
                },
              ],
              "prim": "Pair",
            },
          ],
          "prim": "Pair",
        },
      ],
      "prim": "Pair",
    },
  ],
  "prim": "Pair",
}
`;

exports[`ABI Helpers transforms input to sexp, micheline and raw bytes 3`] = `"050707030a0707009213070700ee5807070100000006737472696e67070700a80f0707010000000761646472657373070701000000036b657907070100000003706b68070701000000097369676e61747572650369"`;
